diff --git a/meson.build b/meson.build
index 790d53d49..00edfe5ed 100644
--- a/meson.build
+++ b/meson.build
@@ -985,6 +985,12 @@ if get_option('newlib-retargetable-locking') != get_option('newlib-multithread')
   error('newlib-retargetable-locking and newlib-multithread must be set to the same value')
 endif
 
+conf_data.set('_GNU_SOURCE', '',
+        description: '''Enable GNU functions like strtof_l.
+It's necessary to set this globally because inline functions in
+libc++ headers call the GNU functions.'''
+)
+
 conf_data.set('_HAVE_CC_INHIBIT_LOOP_TO_LIBCALL',
 	      cc.has_argument('-fno-tree-loop-distribute-patterns'),
 	      description: 'Compiler flag to prevent detecting memcpy/memset patterns')
diff --git a/picolibc.ld.in b/picolibc.ld.in
index f7e7fa80c..75d7a9aaf 100644
--- a/picolibc.ld.in
+++ b/picolibc.ld.in
@@ -207,10 +207,12 @@ SECTIONS
 	 * across them.  We actually need memory allocated for tbss,
 	 * so we create a special segment here just to make room
 	 */
+	/*
 	.tbss_space (NOLOAD) : {
 		. = ADDR(.tbss);
 		. = . + SIZEOF(.tbss);
 	} >ram AT>ram :ram
+	*/
 
 	.bss (NOLOAD) : {
 		*(.sbss*)
diff --git a/test/testcases.c b/test/testcases.c
index 790e782ab..184b8cd9f 100644
--- a/test/testcases.c
+++ b/test/testcases.c
@@ -650,5 +650,8 @@
     /* make sure %c truncates to char */
     wb[0] = 0x34;
     result |= testw(__LINE__, wb, L"%c", wc);
+
+    result |= test(__LINE__, "0.3333333333333333", "%.16Lf", 0.3333333333333333333333333333333333L);
+    result |= test(__LINE__, "3.141592653589793", "%.15Lf", 3.1415926535897932384626433832795029L);
 }
 #endif
diff --git a/test/tls.c b/test/tls.c
index c2ca7e599..bae0e4f3a 100644
--- a/test/tls.c
+++ b/test/tls.c
@@ -43,7 +43,7 @@
 #define DATA_VAL 0x600df00d
 #define DATA_VAL2 0x0a0a0a0a
 
-#define OVERALIGN_DATA  128
+#define OVERALIGN_DATA  256
 #define OVERALIGN_BSS   256
 
 #define TLS_ALIGN      (OVERALIGN_DATA > OVERALIGN_BSS ? OVERALIGN_DATA : OVERALIGN_BSS)
